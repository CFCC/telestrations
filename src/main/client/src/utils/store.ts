import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
