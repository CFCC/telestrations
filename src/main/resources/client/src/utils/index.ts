export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
