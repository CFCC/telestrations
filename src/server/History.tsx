import React, {useState} from "react";
import {Button, MobileStepper} from "@material-ui/core";
import {KeyboardArrowLeft, KeyboardArrowRight} from "@material-ui/icons";
import SwipeableViews from "react-swipeable-views";

interface HistoryProps {
    ownerId?: string;
    playerId?: string;
}

export default function History({ownerId, playerId}: HistoryProps) {
    const [index, setIndex] = useState(0);
    const indexUp = () => setIndex(index + 1);
    const indexDown = () => setIndex(index - 1);

    const pages = [] as any[];

    const nextButton = (
        <Button size="small" onClick={indexUp} disabled={index === pages.length - 1}>
            Next
            <KeyboardArrowRight />
        </Button>
    );
    const backButton = (
        <Button size="small" onClick={indexDown} disabled={index === 0}>
            <KeyboardArrowLeft />
            Back
        </Button>
    );

    return (
        <div>
            <SwipeableViews
                axis="x"
                index={index}
                onChangeIndex={setIndex}
                enableMouseEvents={true}
            >
                {pages.map((page, i) => (
                    <div key={page.text}>
                        {Math.abs(index - i) <= 2 && <img src={atob(page.picture)} alt={page.text} />}
                    </div>
                ))}
            </SwipeableViews>
            <MobileStepper
                steps={pages.length}
                position="static"
                activeStep={index}
                nextButton={nextButton}
                backButton={backButton}
            />
        </div>
    );
}
