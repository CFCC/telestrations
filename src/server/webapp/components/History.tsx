import React, {Component} from 'react';
import {Button, createStyles, MobileStepper, WithStyles} from "@material-ui/core";
import {KeyboardArrowLeft, KeyboardArrowRight} from "@material-ui/icons";
import SwipeableViews from 'react-swipeable-views';
import {State} from "server/webapp/redux/reducers";
import {NotepadPage} from "types/client";
import {UUID} from "types/shared";

const styles = createStyles({

});

const mapStateToProps = (state: State) => ({
    pages: [] as Array<NotepadPage>
});

type HistoryBaseProps = WithStyles<typeof styles> & ReturnType<typeof mapStateToProps>;

interface NotepadHistoryProps {
    notepadOwnerId: UUID;
}

interface PlayerHistoryProps {
    playerId: UUID;
}

type HistoryProps = Partial<HistoryBaseProps> & (NotepadHistoryProps | PlayerHistoryProps);

interface HistoryState {
    index: number;
}

class History extends Component<HistoryProps, HistoryState> {
    state = {
        index: 0
    };

    changeIndex = (index: number) => this.setState({index});

    indexUp = () => this.setState((state: HistoryState) => ({index: state.index + 1}));

    indexDown = () => this.setState((state: HistoryState) => ({index: state.index - 1}));

    render(): React.ReactNode {
        const pages = this.props.pages || [];

        return <div>
            <SwipeableViews
                axis='x'
                index={this.state.index}
                onChangeIndex={this.changeIndex}
                enableMouseEvents
            >
                {pages.map((page: NotepadPage, index: number) => (
                    <div key={page.text}>
                        {Math.abs(this.state.index - index) <= 2 ? (
                            <img src={atob(page.picture)} alt={page.text} />
                        ) : null}
                    </div>
                ))}
            </SwipeableViews>
            <MobileStepper
                steps={pages.length}
                position="static"
                activeStep={this.state.index}
                nextButton={
                    <Button size="small" onClick={this.indexUp} disabled={this.state.index === pages.length - 1}>
                        Next
                        <KeyboardArrowRight />
                    </Button>
                }
                backButton={
                    <Button size="small" onClick={this.indexDown} disabled={this.state.index === 0}>
                        <KeyboardArrowLeft />
                        Back
                    </Button>
                }
            />
        </div>;
    }
}

export default History;
