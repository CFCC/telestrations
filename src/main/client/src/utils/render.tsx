import React from "react";
import { render } from "@testing-library/react";
import { StylesProvider, ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";

import { GlobalStyles, theme } from "./theme";
import { store } from "./store";

interface WrapperProps {
  children: React.ReactNode;
}

const wrapper = ({ children }: WrapperProps) => (
  <React.Fragment>
    <StylesProvider injectFirst={true}>
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </StylesProvider>
  </React.Fragment>
);
const customRender = (ui: React.ReactElement) =>
  render(ui, { wrapper } as object);

export { customRender as render };
