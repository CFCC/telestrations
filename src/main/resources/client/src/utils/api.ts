import SockJS from "sockjs-client";
import { Client } from "@stomp/stompjs";
import _ from "lodash";
import { v4 as uuid } from "uuid";

import { store, actions, rejoinGame } from "./store";
import { Game, Settings } from "./types";

const client = new Client({
  webSocketFactory: () => new SockJS(`${location.origin}/websocket-server`),
});
client.onConnect = () => {
  client.subscribe("/topic/rejoin-game", ({ body }) => {
    store.dispatch(rejoinGame(body));
  });

  client.subscribe("/topic/games", ({ body }) => {
    const response = JSON.parse(body) as string[];
    store.dispatch(actions.handleGamesListMessage(response));
  });

  client.subscribe(`/user/topic/errors/client`, ({ body }) => {
    store.dispatch(actions.handleRequestException(body));
  });

  client.subscribe(`/user/topic/errors/server`, ({ body }) => {
    store.dispatch(actions.handleServerException(body));
  });

  client.subscribe(`/user/topic/successes`, ({ body }) => {
    store.dispatch(actions.handleSuccess(body));
  });
};

export const getGameCodes = (callback: (ids: string[]) => void) => {
  return firebase
    .firestore()
    .collection("games")
    .where("status", "==", "lobby")
    .onSnapshot(async function (snapshot) {
      const newGames = snapshot.docs.map((doc) => doc.id);
      callback(newGames);
    });
};
export const setGameCode = (gameCode: string, isClient: boolean = false) => {
  const game = firebase
    .firestore()
    .doc(`games/${gameCode}`) as DocumentReference<Partial<Game>>;

  game.onSnapshot(async (snapshot) => {
    const newGame = snapshot.data();
    if (!newGame) return;

    if (isClient) {
      if (newGame.status === "in progress") {
        store.dispatch(setGameState(GameState.IN_GAME));
      } else if (newGame.status === "finished") {
        store.dispatch(setGameState(GameState.FINISHED));
        const currentUser = firebase.auth().currentUser;
        if (currentUser?.isAnonymous) {
          await currentUser.delete();
        }
      }
    }

    store.dispatch(updateGame({ ...newGame, id: snapshot.id }));
  });
  game.collection("notepads").onSnapshot((snapshot) => {
    const notepads = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Notepad
    );
    store.dispatch(updateNotepads(notepads));
  });
  game.collection("players").onSnapshot(async (snapshot) => {
    const players = _.mapValues(
      _.keyBy(snapshot.docs, "id"),
      (d) => d.data() as Player
    );
    store.dispatch(updatePlayers(players));

    if (!isClient) {
      const {
        notepads,
        game: { status },
      } = store.getState().firebase;
      const gameIsOver =
        status === "in progress" &&
        Object.entries(players).every(
          ([pid, p]) => notepads[p.currentNotepad]?.ownerId === pid
        ) &&
        Object.values(notepads).every((n) => n.pages.length > 1);

      if (gameIsOver) await game.set({ status: "finished" }, { merge: true });
    } else {
      const { uid } = store.getState().client.user;
      const { currentNotepad, queue } = players[uid];

      if (!currentNotepad && queue?.length > 0) {
        const newQueue = [...queue];
        const newNotepad = newQueue.shift();

        await (firebase
          .firestore()
          .doc(`games/${gameCode}/players/${uid}`) as DocumentReference<
          Partial<Player>
        >).set(
          { queue: newQueue, currentNotepad: newNotepad },
          { merge: true }
        );
        store.dispatch(clientSlice.actions.setGameState(GameState.IN_GAME));
      }
    }
  });
};
export const startGame = async () => {
  const {
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const playerIds = Object.keys(players);

  await Promise.all(
    playerIds.map(async (playerId, i) => {
      const notepadId = uuid();
      await (firebase
        .firestore()
        .doc(
          `games/${gameCode}/notepads/${notepadId}`
        ) as DocumentReference<Notepad>).set({ ownerId: playerId, pages: [] });
      await (firebase
        .firestore()
        .doc(`games/${gameCode}/players/${playerId}`) as DocumentReference<
        Partial<Player>
      >).set(
        {
          currentNotepad: notepadId,
          nextPlayer: playerIds[(i + 1) % playerIds.length],
          queue: [],
        },
        { merge: true }
      );
    })
  );
  await (firebase.firestore().doc(`games/${gameCode}`) as DocumentReference<
    Partial<Game>
  >).set({ status: "in progress" }, { merge: true });
};
export const setGuess = _.throttle(async (guess: string) => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
      notepads,
    },
  } = store.getState();
  const currentNotepadId = players[user.uid].currentNotepad;
  const currentNotepad = notepads[currentNotepadId];
  if (!currentNotepad) return;

  const pages = _.cloneDeep(currentNotepad.pages);
  const needsNewPage = _.last(pages)?.author !== user.uid;
  if (needsNewPage) {
    pages.push({
      author: user.uid,
      lastUpdated: new Date().getTime(),
      content: "",
    });
  }

  if (pages.length % 2 === 1) {
    pages[pages.length - 1].content = guess;
    pages[pages.length - 1].lastUpdated = new Date().getTime();
  } else {
    let fileName = _.last(pages)?.content;
    if (!fileName) {
      fileName = `${uuid()}.png`;
      pages[pages.length - 1].content = fileName;
    }

    pages[pages.length - 1].lastUpdated = new Date().getTime();
    firebase.storage().ref().child(fileName).putString(guess, "data_url");
  }

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/notepads/${currentNotepadId}`) as DocumentReference<
    Partial<Notepad>
  >).set({ pages }, { merge: true });
}, 500);
export const submitGuess = async () => {
  const {
    client: { user },
    firebase: {
      game: { id: gameCode },
      players,
    },
  } = store.getState();
  const { currentNotepad, nextPlayer } = players[user.uid];
  const { queue: nextPlayerQueue } = players[nextPlayer];

  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${nextPlayer}`) as DocumentReference<
    Partial<Player>
  >).set({ queue: [...nextPlayerQueue, currentNotepad] }, { merge: true });
  await (firebase
    .firestore()
    .doc(`games/${gameCode}/players/${user.uid}`) as DocumentReference<
    Partial<Player>
  >).set({ currentNotepad: "" }, { merge: true });
};

export function updateSettings(code: string, settings: Settings) {
  if (!code) return;

  client.publish({
    destination: `/app/games/${code}/update`,
    body: JSON.stringify(settings),
  });
}

export function connectToServer(uuid: string) {
  client.connectHeaders = { uuid };
  client.activate();
}

export function createGame(code: string) {
  client.publish({ destination: `/app/games/${code}/create` });
}

export async function joinGame(
  code: string,
  settings: Settings,
  rejoining: boolean = false
) {
  if (!rejoining) {
    client.publish({
      destination: `/app/games/${code}/join`,
      body: JSON.stringify(settings),
    });
  }

  // A race condition happens where if we dont wait 1ms, the subscription may happen first
  // giving an error that the game doesn't exist
  await new Promise((resolve) => setTimeout(resolve, 1));

  client.subscribe(`/topic/games/${code}`, ({ body }) => {
    const response = JSON.parse(body) as Game;
    store.dispatch(actions.handleGameUpdate(response));
  });
}
