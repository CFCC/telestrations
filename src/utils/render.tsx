import React from "react";
import {render} from "@testing-library/react";
import {StylesProvider, ThemeProvider} from "@material-ui/core/styles";
import {Provider} from "react-redux";

import {GlobalStyles, theme} from "./theme";
import {store} from "../store/client";

interface WrapperProps {
    children: React.ReactNode;
}

const wrapper = (store: any) => ({children}: WrapperProps) => (
    <React.Fragment>
        <StylesProvider injectFirst={true}>
            <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Provider store={store}>
                    {children}
                </Provider>
            </ThemeProvider>
        </StylesProvider>
    </React.Fragment>
);
const customRender = (ui: React.ReactElement, options: any) => render(ui, {wrapper, ...options});

export {customRender as render};
