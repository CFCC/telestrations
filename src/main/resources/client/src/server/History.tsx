import React, { useEffect, useState } from "react";
import { Button, Card, CardMedia, Paper, Typography } from "@material-ui/core";
import { Pagination } from "@material-ui/lab";
import { KeyboardArrowLeft } from "@material-ui/icons";
import SwipeableViews from "react-swipeable-views";
import _ from "lodash";
import styled from "styled-components";
import { useDispatch } from "react-redux";

import { useReduxState } from "../utils/hooks";
import { getImageURL, Page } from "../utils/firebase";
import { clientSlice, GameState } from "../utils/store";
import * as theme from "../utils/theme";

const Header = styled.div`
  background-color: ${theme.secondary};
  color: white;
  width: calc(100% - 2rem);
  height: 2rem;
  padding: 1rem;
  margin-bottom: 2rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  line-height: 2rem !important;
`;

const BackButton = styled(Button)`
  color: white;
  &:hover {
    background-color: rgba(255, 255, 255, 0.04);
  }
`;

const PageStepper = styled(Pagination)`
  button {
    color: white;
  }

  .Mui-selected {
    background-color: rgba(255, 255, 255, 0.08);

    &:hover {
      background-color: rgba(255, 255, 255, 0.12);
    }
  }

  .MuiPaginationItem-page:hover {
    background-color: rgba(255, 255, 255, 0.04);
  }
`;

const Carousel = styled(SwipeableViews)`
  width: 100%;
  flex: 1;

  & > div,
  .slide {
    height: 100%;
  }
`;

const TextContainer = styled(Paper)`
  width: calc(100% - 10rem);
  height: calc(100% - 10rem);
  margin: 5rem;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const ImageContainer = styled(Card)`
  max-width: calc(100% - 10rem);
  max-height: calc(100% - 10rem);
  margin: 5rem;
`;

const Image = styled(CardMedia)`
  padding-top: 56.25%; // 16:9
`;

interface HistoryProps {
  ownerId?: string;
  playerId?: string;
}

export default function History({ ownerId, playerId }: HistoryProps) {
  const [index, setIndex] = useState(0);
  const handleIndexChange = (e: any, v: number) => setIndex(v - 1);

  const {
    firebase: { notepads },
  } = useReduxState();
  let pages: Page[];
  if (ownerId) {
    pages = _.find(notepads, { ownerId })?.pages ?? [];
  } else {
    pages = _.orderBy(
      Object.values(notepads)
        .flatMap((n) => n.pages.map((p, i): [number, Page] => [i, p]))
        .filter((p) => p[1].author === playerId),
      (x) => x[0]
    ).map((x) => x[1]);
  }

  const dispatch = useDispatch();

  function goBack() {
    dispatch(clientSlice.actions.setGameState(GameState.BIRDS_EYE));
  }

  const [content, setContent] = useState(pages);
  useEffect(() => {
    Promise.all(
      pages.map(async (page, i) =>
        i % 2 === 0
          ? page
          : { ...page, content: await getImageURL(page.content) }
      )
    ).then(setContent);
  }, [pages]);

  return (
    <React.Fragment>
      <Header>
        <BackButton onClick={goBack}>
          <KeyboardArrowLeft />
          <Typography>Back to Birds Eye View</Typography>
        </BackButton>
        <Typography variant="h5">
          {ownerId ? "Notepad History" : "Player History"}
        </Typography>
        <PageStepper
          count={pages.length}
          page={index + 1}
          onChange={handleIndexChange}
          size="large"
        />
      </Header>
      <Carousel
        axis="x"
        index={index}
        onChangeIndex={setIndex}
        enableMouseEvents={true}
        slideClassName="slide"
      >
        {content.map((page, i) =>
          i % 2 === 0 ? (
            <TextContainer key={page.lastUpdated}>
              <Typography variant="h1">{page.content}</Typography>
            </TextContainer>
          ) : (
            <ImageContainer key={page.lastUpdated}>
              <Image image={page.content} title="" />
            </ImageContainer>
          )
        )}
      </Carousel>
    </React.Fragment>
  );
}
