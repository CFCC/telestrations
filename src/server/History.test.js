import {render} from "../utils/render";
import {PlayerHistory, NotepadHistory} from "./History.stories";

it('Player History Renders', () => {
    const {asFragment} = render(PlayerHistory());
    expect(asFragment()).toMatchSnapshot();
});

it('Notepad History Renders', () => {
    const {asFragment} = render(NotepadHistory());
    expect(asFragment()).toMatchSnapshot();
});
