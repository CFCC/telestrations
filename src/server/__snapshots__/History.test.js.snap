// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notepad History Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm gFxRDc"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text sc-AxirZ kcEGGr"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Back to Birds Eye View
        </p>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <h5
      class="MuiTypography-root MuiTypography-h5"
    >
      Player History
    </h5>
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root sc-AxiKw ckNwYJ"
    >
      <ul
        class="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <div
    class="sc-AxhCb inUdXy"
    style="overflow-x: hidden;"
  >
    <div
      class="react-swipeable-view-container"
      style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;

exports[`Player History Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm gFxRDc"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text sc-AxirZ kcEGGr"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Back to Birds Eye View
        </p>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <h5
      class="MuiTypography-root MuiTypography-h5"
    >
      Player History
    </h5>
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root sc-AxiKw ckNwYJ"
    >
      <ul
        class="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <div
    class="sc-AxhCb inUdXy"
    style="overflow-x: hidden;"
  >
    <div
      class="react-swipeable-view-container"
      style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
    />
  </div>
</DocumentFragment>
`;
